### EDA - Combined Dateset

In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Reading the dataset
data = pd.read_csv('data/cleaned_data/combined_cleaned.csv')
data.head()

domain  batch  \
0    Web Development Using Python Django  BCK01   
1  Web Development Using Node.js Express  BCK01   
2    Web Development Using Python Django  BCK01   
3                             MERN Stack  BCK01   
4    Web Development Using Python Django  BCK01   

                  placed_company         designation   age     degree cs_bg  \
0           VNC Digital Services   Software Engineer  22.0  No Degree    No   
1  Emstell Technology Consulting   Software Engineer  21.0     Degree    No   
2                 Creative Panda   Node JS Developer  22.0     Degree   Yes   
3              NORQ Technologies  Software Developer  23.0  No Degree    No   
4                       ActionFi   Software Engineer  22.0  No Degree    No   

  experience  avg_time holidays  week_back       city  \
0         No       9.0       No        0.0  Bangalore   
1         No      12.0       No        2.0        WFH   
2         No       9.0      Yes        4.0  Bangalore   
3         No      11.0      Yes        7.0    Chennai   
4         No      10.0       No        4.0  Bangalore   

                                              skills graduated  salary  
0  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.00  
1  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.16  
2  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40  
3  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.55  
4  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40

In [3]:
data.shape

(670, 15)

Dataset contains 670 rows and 15 columns

In [4]:
print(data.columns.values)

['domain' 'batch' 'placed_company' 'designation' 'age' 'degree' 'cs_bg'
 'experience' 'avg_time' 'holidays' 'week_back' 'city' 'skills'
 'graduated' 'salary']


In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 670 entries, 0 to 669
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   domain          670 non-null    object 
 1   batch           670 non-null    object 
 2   placed_company  670 non-null    object 
 3   designation     670 non-null    object 
 4   age             670 non-null    float64
 5   degree          670 non-null    object 
 6   cs_bg           670 non-null    object 
 7   experience      670 non-null    object 
 8   avg_time        670 non-null    float64
 9   holidays        670 non-null    object 
 10  week_back       670 non-null    float64
 11  city            670 non-null    object 
 12  skills          670 non-null    object 
 13  graduated       670 non-null    object 
 14  salary          670 non-null    float64
dtypes: float64(4), object(11)
memory usage: 78.6+ KB


In [6]:
# Dropping column batch
data = data.drop('batch', axis=1)

#### Descriptive Statistics

In [7]:
data.describe().T

count       mean       std   min    25%   50%   75%   max
age        670.0  22.323881  1.242447  19.0  22.00  22.0  23.0  31.0
avg_time   670.0  10.620896  1.985639   7.0   9.25  10.0  12.0  16.0
week_back  670.0   3.395522  2.413275   0.0   1.00   4.0   5.0   7.0
salary     670.0   4.534224  1.693921   1.8   3.60   4.3   5.0  30.0

* Average age of students in Brocamp is 22
* Min age is 19 --> Students joined staright after plus two
* Max age is 31 --> There are people from other professoins who is joining Brocamp
* Average time spent by students is 10.6 hours, which is higher than industrial average (8 hours)
* The least avg time is 7 --> Most possibly day scholars
* Max avg time recorde is 16 --> Students who are putting extra efforts - check corr with salary
* On an average 3 week backs are recieved to the sutdents
* There are students who got zero weekbacks
* Max salary students recieved is 30 LPA - (in UK)
* Min Salary is 1.8 LPA --> May be from early batches
* On an average students get placed with a salary of 4.5 LPA
* The age and salary seems to have right skewed distrubtion (outliers in rightside)

#### Missing values

In [8]:
# Checking for null values
(data.isnull().sum()/len(data))*100

domain            0.0
placed_company    0.0
designation       0.0
age               0.0
degree            0.0
cs_bg             0.0
experience        0.0
avg_time          0.0
holidays          0.0
week_back         0.0
city              0.0
skills            0.0
graduated         0.0
salary            0.0
dtype: float64

In [9]:
# Checking for duplicated values
data.duplicated().sum()

2

The dataset does not have any null values (which were treated in the previous stage) and neither does it have duplicated entries.

#### Data Quality Check

In [10]:
# Function for checking unique values in columsn
def get_unique(column):
    data[column] = data[column].str.lower()
    data[column] = data[column].str.strip()
    unique = data[column].value_counts()
    
    return unique

##### 1. Domain Column

In [11]:
get_unique('domain')

mern stack                                       341
mobile development using flutter                 132
web development using python django               76
web development using python django + react       37
mean stack                                        18
golang                                            14
mobile development using android kotlin            9
machine learning                                   9
web development using node.js express              9
devops                                             8
cybersecurity                                      7
mobile development using swift                     3
game development using unity                       2
data science                                       2
frontend development using react                   1
blockchain                                         1
web development using python django + angular      1
Name: domain, dtype: int64

##### 2. placed_company

In [12]:
data['placed_company'].nunique()

459

* Brocamp has placed students 459 companies
As this is not relevent to our problem statement we can drop this column

In [13]:
data = data.drop('placed_company', axis=1)
data.head()

domain         designation   age     degree  \
0    web development using python django   Software Engineer  22.0  No Degree   
1  web development using node.js express   Software Engineer  21.0     Degree   
2    web development using python django   Node JS Developer  22.0     Degree   
3                             mern stack  Software Developer  23.0  No Degree   
4    web development using python django   Software Engineer  22.0  No Degree   

  cs_bg experience  avg_time holidays  week_back       city  \
0    No         No       9.0       No        0.0  Bangalore   
1    No         No      12.0       No        2.0        WFH   
2   Yes         No       9.0      Yes        4.0  Bangalore   
3    No         No      11.0      Yes        7.0    Chennai   
4    No         No      10.0       No        4.0  Bangalore   

                                              skills graduated  salary  
0  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.00  
1  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.16  
2  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40  
3  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    3.55  
4  Angular, Express, Node, MongoDB, MySQL, Typesc...       Yes    2.40

##### 3. designation 

In [14]:
# Adjnustin pandas settings for printing full values
pd.set_option('display.max_rows', None)

print(get_unique('designation'))

# # Reset the display option to its default value
# pd.reset_option('display.max_rows')

full stack developer                         81
software engineer                            61
flutter developer                            58
software developer                           40
mern stack developer                         36
python developer                             16
associate software engineer                  15
junior software engineer                     13
junior software developer                    13
junior developer                             11
node js developer                            10
junior full stack developer                   9
associate product engineer                    8
junior flutter developer                      8
jr. software engineer                         8
trainee software engineer                     7
backend developer                             6
react developer                               6
front end developer                           6
software engineer trainee                     6
node.js developer                       

* We can see same entries in different formats, may be has extra whitespaces
* Some records are junior, while others has jr.

In [15]:
# Stripping values for removeing extra whitespaces
data['designation'] = data['designation'].str.strip()

print(get_unique('designation'))

full stack developer                         81
software engineer                            61
flutter developer                            58
software developer                           40
mern stack developer                         36
python developer                             16
associate software engineer                  15
junior software engineer                     13
junior software developer                    13
junior developer                             11
node js developer                            10
junior full stack developer                   9
associate product engineer                    8
junior flutter developer                      8
jr. software engineer                         8
trainee software engineer                     7
backend developer                             6
react developer                               6
front end developer                           6
software engineer trainee                     6
node.js developer                       

In [16]:
# Replaceing variations to correct format
replacements = {'jr. software engineer': 'junior software engineer', 'node.js developer ': 'nodejs developer', 'node js developer ': 'nodejs developer',
               'software engineer trainee': 'trainee software engineer', 'software development engineer-1': 'software development engineer',
               'mern developer': 'mern stack developer', 'full stack web developer': 'full stack developer', 'back end developer': 'backend developer',
               'full-stack developer': 'full stack developer', 'mern - stack developer': 'mern stack developer',
               'front-end developer': 'frontend developer', 'front end developer': 'frontend developer',
               'jr. software developer': 'junior software developer', 'sde-1': 'software developer',
               'fullstack developer': 'full stack developer', 'mobile application developer flutter': 'flutter developer',
               'flutter + firebase developer': 'flutter developer ', 'back-end developer': 'backend developer',
               'jr. react developer': 'junior react developer', 'trainee-software engineer': 'trainee software engineer',
               'software development engineer trainee': 'trainee software engineer', 'flutter developer (android/ios)': 'flutter developer',
               'trainee - software developer': 'trainee software developer', 'trainee-software developer': 'trainee software developer', 
               'software engineer(backend)': 'software engineer', 'software developer - trainee': 'trainee software developer',
               'jr.web developer': 'junior web developer', 'jr. react & python developer': 'junior fullstack python developer', 'trainee - software engineer': 'trainee software engineer',
               'oftware development engineer 1': 'software development engineer', 'software engineer (trainee)': 'trainee software engineer',
               'junior node js developer': 'junior nodejs developer', 'jr. app developer ': 'junior app developer',
               'software engineer 1': 'software engineer', 'jr. ml engineer': 'junior ml engineer', 'associate product engineer.': 'associated product engineer',
               'jr. developer': 'junior developer', 'flutter devoloper': 'flutter developer', 'jr. react js developer': 'junior react developer',
               'backend developer (django)': 'bakcend developer', 'fluttter developer': 'flutter developer', 'jr. full stack developer': 'junior full stack developer',
               'junior - full - stack developer': 'junior full stack developer', 'junior full-stack developer': 'junior full stack developer',
               'golang developer(member, technical staff)': 'golang devleoper', 'jr flutter developer': 'junior flutter developer', 'fullstack engineer': 'full stack engineer',
               'jr. flutter developer': 'junior flutter developer', 'junior software engineer(se 1)': 'junior software engineer',
               'golang devleoper': 'golang developer', 'fulll stack developer': 'full stack developer', 'jr node js developer': 'junior nodejs developer',
               'react js developer': 'react developer', 'flutter deevloper': 'flutter developer', 'node.js developer': 'node js developer',
               'jr software engineer': 'junior software engineer', 'bakcend developer': 'backend developer', 'jr. app developer': 'junior app developer',
               'reactjs developer': 'react developer', 'go developer': 'golang developer', 'junior software engineer (android)': 'junior software engineer',
               'jr. machine learning engineer': 'junior machine learning engineer', 'jr full stack python developer': 'junior full stack developer',
               'junior ml engineer': 'junior machine learning engineer', 'node js developer': 'nodejs developer', 'full stack developer': 'fullstack developer',
               'junior full stack developer': 'junior fullstack developer'}

data['designation'] = data['designation'].replace(replacements)
data['designation'].value_counts()

fullstack developer                      81
flutter developer                        64
software engineer                        63
mern stack developer                     41
software developer                       41
junior software engineer                 26
trainee software engineer                18
junior software developer                16
python developer                         16
associate software engineer              15
nodejs developer                         13
junior developer                         12
backend developer                        11
junior flutter developer                 10
full stack developer                      9
junior fullstack developer                9
software development engineer             9
frontend developer                        9
react developer                           8
associate product engineer                8
devops engineer                           5
angular developer                         5
web developer                   

##### 4. city

In [17]:
get_unique('city')

bangalore                    254
wfh                          235
manjeri malappuram            93
chennai                       20
calicut                       14
hyderabad                     11
on site                       11
nagercoil                      8
manjeri                        8
chennai works from office      5
ernakulam                      4
kochi                          2
office                         2
pune                           2
trivandrum                     1
Name: city, dtype: int64

* Bangalore is the city that has more placements, which has simlar number to work from home.
* Some entries as in different format like `manjeri`, and `manjeri malappuarm`
* Most of these data is imputed

In [18]:
replacements = {'manjeri malappuram': 'manjeri', 'on site': 'others', 'chennai works from office': 'chennai',
               'office': 'others'}

data['city'] = data['city'].replace(replacements)
data['city'].value_counts()

bangalore     254
wfh           235
manjeri       101
chennai        25
calicut        14
others         13
hyderabad      11
nagercoil       8
ernakulam       4
kochi           2
pune            2
trivandrum      1
Name: city, dtype: int64

* This is not a reliable columns, as `manjeri` shows high numbers, which is not true - Caused by imputation

#### Extracting uniques `Skills`

In [20]:
# Finding all the skills from the dataset
skills = []

for i in range(len(data)):
    sk = data['skills'][i].split(",")
    
    for j in range(len(sk)):
        skills.append(sk[j])

print(skills)

['Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'Angular', ' Express', ' Node', ' MongoDB', ' MySQL', ' Typescript', ' Html', ' CSS', ' Ajax', ' jquery', ' DataStrures', 'React', ' Node', ' HTM

In [21]:
cleaned_skills = list(set([skill.lower().strip() for skill in skills if skill.strip()]))

print(cleaned_skills)

['api integration', 'vim', 'css', 'nodejs', 'node', 'node js', 'mongodb', 'nest', 'flutter', 'react neative', 'html', 'javascript', 'nextj', 'my sql', 'express js.', 'node.js', 'react. js', 'html css javascript', 'redux', 'computer vision', 'jest', 'tailwind', 'git', 'tailwindcss', 'digital ocean', 'sass', 'go', 'java script', 'python', 'aws', 'deep learning', 'go etc', 'mysql', 'hasura', 'typescipt', 'mongo db', 'nest js', 'react', 'communication', 'next', 'linux', 'niginx', 'jwt', 'react js', 'refine', 'ms office', 'tailwind css', 'socket.io', 'opencv', 'express js', 'express', 'nginx', 'js', 'scss', 'ajax', 'data structure', 'mongoose', 'cnn', 'docker', 'jquery', 'expresses', 'firebase', 'mui etc..', 'sql', 'postman', 'typescript', 'dart', 'django', 'jquary', 'datastrures', 'html& css', 'socket io', 'vercel', 'angular', 'react query', 'firbase', 'graphql', 'api', 'mui', 'bootstrap']


In [22]:
len(cleaned_skills)

80

In [25]:
import nltk
from nltk.stem import WordNetLemmatizer
import re

# Standardize skill names
wordnet_lemmatizer = WordNetLemmatizer()
unique_skills = set()
for skill in cleaned_skills:
    # Convert to lowercase and remove punctuation and white spaces
    skill = skill.lower().replace(".", "").replace(",", "").replace(" ", "")
    # Lemmatize skill name
    skill = wordnet_lemmatizer.lemmatize(skill)
    unique_skills.add(skill)

# Print unique skill names
print(sorted(unique_skills))

['ajax', 'angular', 'api', 'apiintegration', 'aws', 'bootstrap', 'cnn', 'communication', 'computervision', 'cs', 'dart', 'datastructure', 'datastrures', 'deeplearning', 'digitalocean', 'django', 'docker', 'express', 'expressjs', 'firbase', 'firebase', 'flutter', 'git', 'go', 'goetc', 'graphql', 'hasura', 'html', 'html&css', 'htmlcssjavascript', 'j', 'javascript', 'jest', 'jquary', 'jquery', 'jwt', 'linux', 'mongodb', 'mongoose', 'msoffice', 'mui', 'muietc', 'mysql', 'nest', 'nestjs', 'next', 'nextj', 'nginx', 'niginx', 'node', 'nodejs', 'opencv', 'postman', 'python', 'react', 'reactjs', 'reactneative', 'reactquery', 'redux', 'refine', 'sas', 'sc', 'socketio', 'sql', 'tailwind', 'tailwindcss', 'typescipt', 'typescript', 'vercel', 'vim']


In [24]:
len(unique_skills)

70

In [27]:
unique_skills = sorted(list(unique_skills))

replacements = {
    'datastrures': 'datastructures',
    'datastructure': 'datastructures',
    'express': 'expressjs',
    'firbase': 'firebase',
    'html&css': 'html',
    'go': 'mongodb',
    'apiintegration': 'api',
    'nextj': 'nextjs',
    'sc': 'scss',
    'tailwind': 'tailwindcss',
    'goetc': 'go',
    'niginx': 'nginx',
    'react': 'reactjs',
    'node': 'nodejs',
    'refine': 'refinejs',
    'nextj': 'nextjs',
    'j': 'javascript',
    'reactquery': 'reactjs',
    'muietc': 'mui',
    'htmlcssjavascript': 'javascript',
    'jquary': 'jquery',
    'cs': 'css',
    'nest': 'nestjs',
    'reactneative': 'reactnative',
    'typescipt': 'typescript'
}

unique_skills_cleaned = set()

# Iterate over the skills list and extract unique skills
for skill in unique_skills:
    # Split the skill string on space and/or dot characters
    parts = re.split(r'[.\s]+', skill.lower())
    # Iterate over the parts and add non-empty and non-stopword parts to the unique_skills set
    for part in parts:
        if part not in replacements and part not in nltk.corpus.stopwords.words('english') and len(part) > 1:
            unique_skills_cleaned.add(part)

# Replace any keywords in the unique skills with their replacements
unique_skills_cleaned = [replacements.get(skill, skill) for skill in unique_skills_cleaned]

# Sort the unique skills list and print it
print(sorted(unique_skills_cleaned))

['ajax', 'angular', 'api', 'aws', 'bootstrap', 'cnn', 'communication', 'computervision', 'dart', 'deeplearning', 'digitalocean', 'django', 'docker', 'expressjs', 'firebase', 'flutter', 'git', 'graphql', 'hasura', 'html', 'javascript', 'jest', 'jquery', 'jwt', 'linux', 'mongodb', 'mongoose', 'msoffice', 'mui', 'mysql', 'nestjs', 'next', 'nginx', 'nodejs', 'opencv', 'postman', 'python', 'reactjs', 'redux', 'sas', 'socketio', 'sql', 'tailwindcss', 'typescript', 'vercel', 'vim']


In [28]:
len(unique_skills_cleaned)

46

##### Creating columns for each unique skill

In [29]:
# function to extract unique skills from the skills column
def extract_skills(skills):
    extracted_skills = []
    for skill in unique_skills_cleaned:
        pattern = re.compile(skill, re.IGNORECASE)
        if any(pattern.search(s) for s in skills.split(',')):
            extracted_skills.append(skill)
    return extracted_skills

In [57]:
# Creating a copy of the dataframe
skills_df = data.copy()

# extract unique skills from skills column
skills_df['extracted_skills'] = skills_df['skills'].apply(extract_skills)

# create new columns for each unique skill
for skill in unique_skills_cleaned:
    skills_df[skill] = skills_df['extracted_skills'].apply(lambda x: 1 if skill in x else 0)

# drop the extracted_skills column
skills_df = skills_df.drop('extracted_skills', axis=1)

skills_df.head()

domain         designation   age     degree  \
0    web development using python django   software engineer  22.0  No Degree   
1  web development using node.js express   software engineer  21.0     Degree   
2    web development using python django    nodejs developer  22.0     Degree   
3                             mern stack  software developer  23.0  No Degree   
4    web development using python django   software engineer  22.0  No Degree   

  cs_bg experience  avg_time holidays  week_back       city  ... linux next  \
0    No         No       9.0       No        0.0  bangalore  ...     0    0   
1    No         No      12.0       No        2.0        wfh  ...     0    0   
2   Yes         No       9.0      Yes        4.0  bangalore  ...     0    0   
3    No         No      11.0      Yes        7.0    chennai  ...     0    0   
4    No         No      10.0       No        4.0  bangalore  ...     0    0   

   tailwindcss  deeplearning  mui  jwt  jquery  bootstrap  reactjs  firebase  
0            0             0    0    0       1          0        0         0  
1            0             0    0    0       1          0        0         0  
2            0             0    0    0       1          0        0         0  
3            0             0    0    0       1          0        0         0  
4            0             0    0    0       1          0        0         0  

[5 rows x 59 columns]

In [58]:
skills_df.columns

Index(['domain', 'designation', 'age', 'degree', 'cs_bg', 'experience',
       'avg_time', 'holidays', 'week_back', 'city', 'skills', 'graduated',
       'salary', 'python', 'vim', 'nodejs', 'sql', 'postman', 'sas', 'opencv',
       'aws', 'typescript', 'computervision', 'dart', 'mongodb', 'nestjs',
       'flutter', 'django', 'nginx', 'html', 'javascript', 'hasura', 'mysql',
       'ajax', 'expressjs', 'vercel', 'msoffice', 'redux', 'socketio',
       'digitalocean', 'mongoose', 'cnn', 'angular', 'jest', 'git', 'graphql',
       'docker', 'api', 'communication', 'linux', 'next', 'tailwindcss',
       'deeplearning', 'mui', 'jwt', 'jquery', 'bootstrap', 'reactjs',
       'firebase'],
      dtype='object')

In [59]:
# Moving the `salary` column to last
column_to_move = skills_df.pop("salary")

# insert column with insert(location, column_name, column_value)

skills_df.insert(58, "salary", column_to_move)

In [60]:
# Dropping the skills column
skills_df = skills_df.drop('skills', axis=1)

In [61]:
skills_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 670 entries, 0 to 669
Data columns (total 58 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   domain          670 non-null    object 
 1   designation     670 non-null    object 
 2   age             670 non-null    float64
 3   degree          670 non-null    object 
 4   cs_bg           670 non-null    object 
 5   experience      670 non-null    object 
 6   avg_time        670 non-null    float64
 7   holidays        670 non-null    object 
 8   week_back       670 non-null    float64
 9   city            670 non-null    object 
 10  graduated       670 non-null    object 
 11  python          670 non-null    int64  
 12  vim             670 non-null    int64  
 13  nodejs          670 non-null    int64  
 14  sql             670 non-null    int64  
 15  postman         670 non-null    int64  
 16  sas             670 non-null    int64  
 17  opencv          670 non-null    int

In [62]:
skills_df.describe().T

count       mean       std   min    25%   50%   75%   max
age             670.0  22.323881  1.242447  19.0  22.00  22.0  23.0  31.0
avg_time        670.0  10.620896  1.985639   7.0   9.25  10.0  12.0  16.0
week_back       670.0   3.395522  2.413275   0.0   1.00   4.0   5.0   7.0
python          670.0   0.035821  0.185982   0.0   0.00   0.0   0.0   1.0
vim             670.0   0.014925  0.121345   0.0   0.00   0.0   0.0   1.0
nodejs          670.0   0.002985  0.054595   0.0   0.00   0.0   0.0   1.0
sql             670.0   0.691045  0.462408   0.0   0.00   1.0   1.0   1.0
postman         670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
sas             670.0   0.023881  0.152791   0.0   0.00   0.0   0.0   1.0
opencv          670.0   0.022388  0.148053   0.0   0.00   0.0   0.0   1.0
aws             670.0   0.371642  0.483604   0.0   0.00   0.0   1.0   1.0
typescript      670.0   0.443284  0.497144   0.0   0.00   0.0   1.0   1.0
computervision  670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
dart            670.0   0.025373  0.157373   0.0   0.00   0.0   0.0   1.0
mongodb         670.0   0.829851  0.376045   0.0   1.00   1.0   1.0   1.0
nestjs          670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
flutter         670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
django          670.0   0.013433  0.115205   0.0   0.00   0.0   0.0   1.0
nginx           670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
html            670.0   0.947761  0.222674   0.0   1.00   1.0   1.0   1.0
javascript      670.0   0.016418  0.127171   0.0   0.00   0.0   0.0   1.0
hasura          670.0   0.001493  0.038633   0.0   0.00   0.0   0.0   1.0
mysql           670.0   0.338806  0.473657   0.0   0.00   0.0   1.0   1.0
ajax            670.0   0.319403  0.466594   0.0   0.00   0.0   1.0   1.0
expressjs       670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
vercel          670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
msoffice        670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
redux           670.0   0.035821  0.185982   0.0   0.00   0.0   0.0   1.0
socketio        670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
digitalocean    670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
mongoose        670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
cnn             670.0   0.022388  0.148053   0.0   0.00   0.0   0.0   1.0
angular         670.0   0.334328  0.472108   0.0   0.00   0.0   1.0   1.0
jest            670.0   0.008955  0.094278   0.0   0.00   0.0   0.0   1.0
git             670.0   0.329851  0.470510   0.0   0.00   0.0   1.0   1.0
graphql         670.0   0.001493  0.038633   0.0   0.00   0.0   0.0   1.0
docker          670.0   0.038806  0.193277   0.0   0.00   0.0   0.0   1.0
api             670.0   0.029851  0.170303   0.0   0.00   0.0   0.0   1.0
communication   670.0   0.001493  0.038633   0.0   0.00   0.0   0.0   1.0
linux           670.0   0.014925  0.121345   0.0   0.00   0.0   0.0   1.0
next            670.0   0.010448  0.101755   0.0   0.00   0.0   0.0   1.0
tailwindcss     670.0   0.307463  0.461787   0.0   0.00   0.0   1.0   1.0
deeplearning    670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
mui             670.0   0.029851  0.170303   0.0   0.00   0.0   0.0   1.0
jwt             670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
jquery          670.0   0.350746  0.477560   0.0   0.00   0.0   1.0   1.0
bootstrap       670.0   0.046269  0.210223   0.0   0.00   0.0   0.0   1.0
reactjs         670.0   0.000000  0.000000   0.0   0.00   0.0   0.0   0.0
firebase        670.0   0.004478  0.066815   0.0   0.00   0.0   0.0   1.0
salary          670.0   4.534224  1.693921   1.8   3.60   4.3   5.0  30.0

### Univariate analysis

#### Distributions of Features and Target